<h1 align=center><font size=5>Data Analysis with Python</font></h1>

<h1>Data Preprocessing Assignment</h1>

<h3>Individuale Assignment!</h3>

Dear all, Please complete this assignment and submit to google drive

<h4>Import pandas</h4> 

In [96]:
# Write your code below 
import pandas as pd
import numpy as np


Use the Pandas method <b>read_csv()</b> to load the data (mydataset). The dataset is available in the same drive

In [97]:
# Write your code below "mydataset.csv"
df= pd.read_csv('mydataset.csv')
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg car-price  
0          27     13495  
1          27     16500  
2          26     16500  
3          30     13950  
4          22     17450  

[5 rows x 26 columns]

 Use the method <b>head()</b> to display the first five rows of the dataframe. 

In [98]:
# To see what the data set looks like, we'll use the head() method.
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg car-price  
0          27     13495  
1          27     16500  
2          26     16500  
3          30     13950  
4          22     17450  

[5 rows x 26 columns]

<h2 id="identify_handle_missing_values">Identify and handle missing values</h2>


<h3 id="identify_missing_values">Identify missing values</h3>
<h4>Convert "?" to NaN</h4>
In the car dataset, missing data comes with the question mark "?".
replace "?" with NaN (Not a Number), which is Python's default missing value marker, for reasons of computational speed and convenience. Here we use the function: 
 <pre>.replace(A, B, inplace = True) </pre>
to replace A by B

In [99]:
# Write your code below 
data =pd.DataFrame(df)
data.replace("?", np.nan, inplace=True)
data.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg car-price  
0          27     13495  
1          27     16500  
2          26     16500  
3          30     13950  
4          22     17450  

[5 rows x 26 columns]

<h4>Evaluating for Missing Data</h4>

The missing values are converted to Python's default. We use Python's built-in functions to identify these missing values. There are two methods to detect missing data:
<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.

In [100]:
# Write your code below 
df.isnull()
df.notnull()

symboling  normalized-losses  make  fuel-type  aspiration  num-of-doors  \
0         True              False  True       True        True          True   
1         True              False  True       True        True          True   
2         True              False  True       True        True          True   
3         True               True  True       True        True          True   
4         True               True  True       True        True          True   
..         ...                ...   ...        ...         ...           ...   
200       True               True  True       True        True          True   
201       True               True  True       True        True          True   
202       True               True  True       True        True          True   
203       True               True  True       True        True          True   
204       True               True  True       True        True          True   

     body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0          True          True             True        True  ...         True   
1          True          True             True        True  ...         True   
2          True          True             True        True  ...         True   
3          True          True             True        True  ...         True   
4          True          True             True        True  ...         True   
..          ...           ...              ...         ...  ...          ...   
200        True          True             True        True  ...         True   
201        True          True             True        True  ...         True   
202        True          True             True        True  ...         True   
203        True          True             True        True  ...         True   
204        True          True             True        True  ...         True   

     fuel-system  bore  stroke  compression-ratio  horsepower  peak-rpm  \
0           True  True    True               True        True      True   
1           True  True    True               True        True      True   
2           True  True    True               True        True      True   
3           True  True    True               True        True      True   
4           True  True    True               True        True      True   
..           ...   ...     ...                ...         ...       ...   
200         True  True    True               True        True      True   
201         True  True    True               True        True      True   
202         True  True    True               True        True      True   
203         True  True    True               True        True      True   
204         True  True    True               True        True      True   

     city-mpg  highway-mpg  car-price  
0        True         True       True  
1        True         True       True  
2        True         True       True  
3        True         True       True  
4        True         True       True  
..        ...          ...        ...  
200      True         True       True  
201      True         True       True  
202      True         True       True  
203      True         True       True  
204      True         True       True  

[205 rows x 26 columns]

"True" stands for missing value, while "False" stands for not missing value.

<h4>Count missing values in each column</h4>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value, "False"  means the value is present in the dataset.  In the body of the for loop the method  ".value_counts()"  counts the number of "True" values. 
</p>

In [101]:
# Write your code below 

for column in df.columns:
    missing_count = df[column].isnull().sum()
    print(f"Column '{column}' has {missing_count} missing values.")


Column 'symboling' has 0 missing values.
Column 'normalized-losses' has 41 missing values.
Column 'make' has 0 missing values.
Column 'fuel-type' has 0 missing values.
Column 'aspiration' has 0 missing values.
Column 'num-of-doors' has 2 missing values.
Column 'body-style' has 0 missing values.
Column 'drive-wheels' has 0 missing values.
Column 'engine-location' has 0 missing values.
Column 'wheel-base' has 0 missing values.
Column 'length' has 0 missing values.
Column 'width' has 0 missing values.
Column 'height' has 0 missing values.
Column 'curb-weight' has 0 missing values.
Column 'engine-type' has 0 missing values.
Column 'num-of-cylinders' has 0 missing values.
Column 'engine-size' has 0 missing values.
Column 'fuel-system' has 0 missing values.
Column 'bore' has 4 missing values.
Column 'stroke' has 4 missing values.
Column 'compression-ratio' has 0 missing values.
Column 'horsepower' has 2 missing values.
Column 'peak-rpm' has 2 missing values.
Column 'city-mpg' has 0 missing v

Based on the summary above, each column has 205 rows of data, seven columns containing missing data:
<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"car-price": 4 missing data</li>
</ol>

<h3 id="deal_missing_values">Deal with missing data</h3>
<b>How to deal with missing data?</b>

<ol>
    <li>drop data<br>
        a. drop the whole row<br>
        b. drop the whole column
    </li>
    <li>replace data<br>
        a. replace it by mean<br>
        b. replace it by frequency<br>
        c. replace it based on other functions
    </li>
</ol>

<b>Replace by mean:</b>
<ul>
    <li>"normalized-losses": 41 missing data, replace them with mean</li>
    <li>"stroke": 4 missing data, replace them with mean</li>
    <li>"bore": 4 missing data, replace them with mean</li>
    <li>"horsepower": 2 missing data, replace them with mean</li>
    <li>"peak-rpm": 2 missing data, replace them with mean</li>
</ul>

<b>Replace by frequency:</b>
<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four". 
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

<b>Drop the whole row:</b>
<ul>
    <li>"car-price": 4 missing data, simply delete the whole row
        <ul>
            <li>Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us</li>
        </ul>
    </li>
</ul>

<h4>Calculate the mean value for "normalized-losses" </h4>

In [102]:
mean_normalized_losses = df['normalized-losses'].astype(float).mean()

mean_normalized_losses

122.0

<h4>Replace "NaN" by mean value in "normalized-losses" column</h4>

In [103]:
# Write your code below 

df['normalized-losses'].fillna(mean_normalized_losses, inplace=True)

df['normalized-losses']

C:\Users\user\AppData\Local\Temp\ipykernel_7956\2552521859.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['normalized-losses'].fillna(mean_normalized_losses, inplace=True)


0      122.0
1      122.0
2      122.0
3        164
4        164
       ...  
200       95
201       95
202       95
203       95
204       95
Name: normalized-losses, Length: 205, dtype: object

<h4>Calculate the mean value for 'bore' column</h4>

In [104]:
# Calculate the mean of the bore
df['bore'] = df['bore'].astype(float)

mean_bore = df['bore'].mean()
print(mean_bore)

3.3297512437810943


<h4>Replace NaN by mean value</h4>

In [105]:
# Write your code below 
df['bore'].fillna(mean_bore, inplace=True)

print(df)

     symboling normalized-losses         make fuel-type aspiration  \
0            3             122.0  alfa-romero       gas        std   
1            3             122.0  alfa-romero       gas        std   
2            1             122.0  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1

C:\Users\user\AppData\Local\Temp\ipykernel_7956\3643831839.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['bore'].fillna(mean_bore, inplace=True)


<b>According to the example above, replace NaN in "stroke" column by mean.</b>

In [106]:
# Write your code below 
df['stroke'].fillna('NaN', inplace=True)

print(df)

     symboling normalized-losses         make fuel-type aspiration  \
0            3             122.0  alfa-romero       gas        std   
1            3             122.0  alfa-romero       gas        std   
2            1             122.0  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1

C:\Users\user\AppData\Local\Temp\ipykernel_7956\1325789522.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['stroke'].fillna('NaN', inplace=True)


<h4>Calculate the mean value for the  'horsepower' column:</h4>

In [107]:
# Write your code below 

df['horsepower'] = pd.to_numeric(df['horsepower'], errors='coerce')

mean_horsepower = df['horsepower'].mean()

print(f"Mean of horsepower: {mean_horsepower}")


Mean of horsepower: 104.25615763546799


<h4>Replace "NaN" by mean value:</h4>

In [108]:
# Write your code below 
df['horsepower'].fillna(mean_horsepower, inplace=True)

print(df)

     symboling normalized-losses         make fuel-type aspiration  \
0            3             122.0  alfa-romero       gas        std   
1            3             122.0  alfa-romero       gas        std   
2            1             122.0  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1

C:\Users\user\AppData\Local\Temp\ipykernel_7956\2800120259.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['horsepower'].fillna(mean_horsepower, inplace=True)


<h4>Calculate the mean value for 'peak-rpm' column:</h4>

In [109]:
# Write your code below 
df['peak-rpm'] = pd.to_numeric(df['peak-rpm'], errors='coerce')

mean_peak_rpm = df['peak-rpm'].mean()

print(f"Mean of peak-rpm: {mean_peak_rpm}")


Mean of peak-rpm: 5125.369458128079


<h4>Replace NaN by mean value:</h4>

In [110]:
# Write your code below 
df['peak-rpm'].fillna(mean_peak_rpm, inplace=True)

print(df)

     symboling normalized-losses         make fuel-type aspiration  \
0            3             122.0  alfa-romero       gas        std   
1            3             122.0  alfa-romero       gas        std   
2            1             122.0  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1

C:\Users\user\AppData\Local\Temp\ipykernel_7956\136821064.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['peak-rpm'].fillna(mean_peak_rpm, inplace=True)


<b>Replace by frequency:</b>
<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four". 
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

To see which values are present in a particular column, please use the ".value_counts()" method:

In [111]:
# Write your code below for "num-of-doors"
door_counts = df['num-of-doors'].value_counts()

df['num-of-doors'].fillna('four', inplace=True)

print(df['num-of-doors'])

0       two
1       two
2       two
3      four
4      four
       ... 
200    four
201    four
202    four
203    four
204    four
Name: num-of-doors, Length: 205, dtype: object


C:\Users\user\AppData\Local\Temp\ipykernel_7956\978486812.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['num-of-doors'].fillna('four', inplace=True)


As you can see that four doors are the most common type. You can also use the ".idxmax()" method to calculate for us the most common type automatically:

In [112]:
# Write your code below 
most_frequent_door = df['num-of-doors'].value_counts().idxmax()

df['num-of-doors'].fillna(most_frequent_door, inplace=True)

print(df['num-of-doors'])

0       two
1       two
2       two
3      four
4      four
       ... 
200    four
201    four
202    four
203    four
204    four
Name: num-of-doors, Length: 205, dtype: object


The replacement procedure is very similar to what we have seen previously

In [113]:
most_frequent_door = df['num-of-doors'].value_counts().idxmax()

df['num-of-doors'].fillna(most_frequent_door, inplace=True)

print(df['num-of-doors'])

0       two
1       two
2       two
3      four
4      four
       ... 
200    four
201    four
202    four
203    four
204    four
Name: num-of-doors, Length: 205, dtype: object


<b>Drop the whole row:</b>
<ul>
    <li>"car-price": 4 missing data, simply delete the whole row
        <ul>
            <li>Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us</li>
        </ul>
    </li>
</ul>

Finally, let's drop all rows that do not have price data:

In [114]:
# simply drop whole row with NaN in "car-price" column
df.dropna(subset=["car-price"], axis=0, inplace=True)
df.reset_index(drop=True, inplace=True)
print(df.head())

   symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio hor

<b>Good!</b> Now, we obtain the dataset with no missing values.

<h3 id="correct_data_format">Correct data format</h3>
<p>The last step in data cleaning is checking and making sure that all data is in the correct format (int, float, text or other).</p>

In Pandas, we use 
<p><b>.dtype()</b> to check the data type</p>
<p><b>.astype()</b> to change the data type</p>

<h4>Lets list the data types for each column</h4>

In [115]:
# Your code here
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke                object
compression-ratio    float64
horsepower           float64
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
car-price             object
dtype: object


As we can see above, some columns are not of the correct data type. Numerical variables should have type 'float' or 'int', and variables with strings such as categories should have type 'object'. For example, **'bore'** and **'stroke'** variables are numerical values that describe the engines, so we should expect them to be of the type *'float'* or *'int'*; however, they are shown as type 'object'. We have to convert data types into a proper format for each column using the "astype()" method.

<h4>Convert data types to proper format</h4>

In [116]:
df['bore'] = pd.to_numeric(df['bore'])  
df['stroke'] = pd.to_numeric(df['stroke'])
df['horsepower'] = pd.to_numeric(df['horsepower'])
df['peak-rpm'] = pd.to_numeric(df['peak-rpm'])
df['car-price'] = pd.to_numeric(df['car-price'])
df['num-of-doors'] = df['num-of-doors'].astype(str)


ValueError: Unable to parse string "NaN" at position 52

<h4>List the columns after the conversion</h4>

In [117]:
# Write your code below 
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke                object
compression-ratio    float64
horsepower           float64
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
car-price             object
dtype: object


<b>Wonderful!</b>

Now, we finally obtain the cleaned dataset with no missing values and all data in its proper format.

<h2 id="indicator">Indicator variable (or dummy variable)</h2>
<p>
    An indicator variable (or dummy variable) is a numerical variable used to label categories. They are called 'dummies' because the numbers themselves don't have inherent meaning. 
</p><br>
Example

We see the column **fuel-type** has two unique values, "gas" or "diesel". Regression doesn't understand words, only numbers. To use this attribute in regression analysis, we convert "fuel-type" into indicator variables.


<p>
    We will use the panda's method 'get_dummies' to assign numerical values to different categories of fuel type. 
</p>

get indicator variables and assign it to data frame "dummy_variable" 

In [118]:
# Write your code below for "fuel-type" feature. 
dummy_variable = pd.get_dummies(df['fuel-type'], prefix='fuel-type')

df = pd.concat([df, dummy_variable], axis=1)

print(df.head())

   symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  bore  stroke  \
0  convertible          rwd           front        88.6  ...  3.47    2.68   
1  convertible          rwd           front        88.6  ...  3.47    2.68   
2    hatchback          rwd           front        94.5  ...  2.68    3.47   
3        sedan          fwd           front        99.8  ...  3.19    3.40   
4        sedan          4wd           front        99.4  ...  3.19    3.40   

   compression-ratio  horsepower peak-rpm ci

change column names for clarity 

In [119]:
# Assuming your DataFrame has 30 columns, modify the list accordingly
df.columns = ['Symboling', 'Normalized_Losses', 'Make', 'Fuel_Type', 'Aspiration', 
              'Num_Of_Doors', 'Body_Style', 'Drive_Wheels', 'Engine_Location', 
              'Wheel_Base', 'Length', 'Width', 'Height', 'Curb_Weight', 'Engine_Type', 
              'Num_Of_Cylinders', 'Engine_Size', 'Fuel_System', 'Bore', 'Stroke', 
              'Compression_Ratio', 'Horsepower', 'Peak_RPM', 'City_MPG', 'Highway_MPG', 
              'Car_Price', 'Fuel_Type_Gas', 'Fuel_Type_Diesel', 'Aspiration_Std', 'Aspiration_Turbo']

# Display the updated column names
print(df.columns)


ValueError: Length mismatch: Expected axis has 28 elements, new values have 30 elements

We now have the value 0 to represent "gas" and 1 to represent "diesel" in the column "fuel-type". We will now insert this column back into our original dataset. 

In [120]:

dummyOne = pd.get_dummies(df['fuel-type'], prefix='fuel-type')

df = pd.concat([df, dummyOne], axis=1)

dummyOne.head()
y=df['fuel-type']


df=pd.concat([df,dummyOne],axis=1)

df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  peak-rpm  \
0  convertible          rwd           front        88.6  ...    5000.0   
1  convertible          rwd           front        88.6  ...    5000.0   
2    hatchback          rwd           front        94.5  ...    5000.0   
3        sedan          fwd           front        99.8  ...    5500.0   
4        sedan          4wd           front        99.4  ...    5500.0   

   city-mpg  highway-mpg  car-price  fuel-type_diesel fuel-type_gas  \
0        21           27      13495             False          True   
1        21           27      16500             False          True   
2        19           26      16500             False          True   
3        24           30      13950             False          True   
4        18           22      17450             False          True   

   fuel-type_diesel fuel-type_gas  fuel-type_diesel fuel-type_gas  
0             False          True             False          True  
1             False          True             False          True  
2             False          True             False          True  
3             False          True             False          True  
4             False          True             False          True  

[5 rows x 32 columns]

The last two columns are now the indicator variable representation of the fuel-type variable. It's all 0s and 1s now.

<b>As above, create indicator variable to the column of "aspiration": "std" to 0, while "turbo" to 1.</b>

In [121]:
# Keep only the 'Aspiration_turbo' column as the indicator variable
aspiration_dummies = aspiration_dummies['Aspiration_turbo']

print(aspiration_dummies.head())

NameError: name 'aspiration_dummies' is not defined

<b>Merge the new dataframe to the original dataframe then drop the column 'aspiration'</b>

In [122]:
# Check if 'aspiration' column exists before dropping it
if 'aspiration' in df.columns:
    df.drop('aspiration', axis=1, inplace=True)
else:
    print("The 'aspiration' column is already removed or does not exist.")

print(df.head())


   symboling normalized-losses         make fuel-type num-of-doors  \
0          3             122.0  alfa-romero       gas          two   
1          3             122.0  alfa-romero       gas          two   
2          1             122.0  alfa-romero       gas          two   
3          2               164         audi       gas         four   
4          2               164         audi       gas         four   

    body-style drive-wheels engine-location  wheel-base  length  ...  \
0  convertible          rwd           front        88.6   168.8  ...   
1  convertible          rwd           front        88.6   168.8  ...   
2    hatchback          rwd           front        94.5   171.2  ...   
3        sedan          fwd           front        99.8   176.6  ...   
4        sedan          4wd           front        99.4   176.6  ...   

   peak-rpm  city-mpg  highway-mpg car-price fuel-type_diesel  fuel-type_gas  \
0    5000.0        21           27     13495            False     

**save your file and upload to the Google Drive.** 

Source: IBM, MIT